drivers/usb_c/tcpc/ucpd_numaker.c

```diff
@@ -0,0 +1,2528 @@
+
+/**
+ * @brief Immutable device context
+ */
+struct numaker_tcpc_config {
+	UTCPD_T *utcpd_base;
+	EADC_T *eadc_base;
+	TIMER_T *timer_base;
+	const struct device *clkctrl_dev;
+	struct numaker_scc_subsys pcc_utcpd;
+	struct numaker_scc_subsys pcc_timer;
+	struct reset_dt_spec reset_utcpd;
+	struct reset_dt_spec reset_timer;
+	void (*irq_config_func_utcpd)(const struct device *dev);
+	void (*irq_unconfig_func_utcpd)(const struct device *dev);
+	const struct pinctrl_dev_config *pincfg;
+	struct {
+		struct {
+			struct gpio_dt_spec vbus_detect;
+			struct gpio_dt_spec vbus_discharge;
+			struct gpio_dt_spec vconn_discharge;
+		} gpios;
+
+		bool dead_battery;
+		struct {
+			uint32_t bit;
+		} pinpl;
+		struct {
+			struct {
+				uint32_t bit;
+				uint32_t value;
+			} vbscale;
+		} vbvol;
+	} utcpd;
+	struct {
+		const struct adc_dt_spec *spec_vbus;
+		const struct adc_dt_spec *spec_vconn;
+		/* Rate of timer-triggered voltage measurement (Hz) */
+		uint32_t timer_trigger_rate;
+		/* Trigger source for measuring VBUS/VCONN voltage */
+		uint32_t trgsel_vbus;
+		uint32_t trgsel_vconn;
+	} eadc;
+};
+
+/**
+ * @brief Mutable device context
+ */
+struct numaker_tcpc_data {
+	enum tc_rp_value rp;
+	bool rx_sop_prime_enabled;
+
+	/* One-slot Rx FIFO */
+	bool rx_msg_ready;
+	struct pd_msg rx_msg;
+
+	/* Mark the end of zero-initialized fields in tcpc_init(). The
+	 * fields following the marker will persist across tcpc_init().
+	 */
+	uint32_t zero_end_marker;
+
+	uint32_t vref_mv;
+
+	/* TCPC alert */
+	struct {
+		tcpc_alert_handler_cb_t handler;
+		void *data;
+	} tcpc_alert;
+
+	/* PPC event */
+	struct {
+		usbc_ppc_event_cb_t handler;
+		void *data;
+	} ppc_event;
+
+	/* DPM supplied */
+	struct {
+		/* VCONN callback function */
+		tcpc_vconn_control_cb_t vconn_cb;
+		/* VCONN discharge callback function */
+		tcpc_vconn_discharge_cb_t vconn_discharge_cb;
+	} dpm;
+};
+
+/**
+ * @brief Wait ready for next write access to UTCPD register
+ *
+ * @retval 0 on success
+ * @retval -EIO on failure
+ */
+static int numaker_utcpd_reg_write_wait_ready(const struct device *dev)
+{
+	const struct numaker_tcpc_config *const config = dev->config;
+	UTCPD_T *utcpd_base = config->utcpd_base;
+
+	if (!WAIT_FOR((utcpd_base->CLKINFO & UTCPD_CLKINFO_ReadyFlag_Msk),
+		      NUMAKER_UTCPD_REG_WRITE_BY_NAME_TIMEOUT_US, NULL)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Convert VBUS voltage format from H/W bit to mV
+ *
+ * The following factors are taken into consideration:
+ * 1. Analog Vref
+ * 2. UTCPD VBVOL.VBSCALE
+ *
+ * @note UTCPD VBVOL.VBVOL = MSB 10-bit of EADC DAT.RESULT[11:0],
+ *       that is, discarding LSB 2-bit.
+ */
+static uint32_t numaker_utcpd_vbus_volt_bit2mv(const struct device *dev, uint32_t bit)
+{
+	const struct numaker_tcpc_config *const config = dev->config;
+	struct numaker_tcpc_data *data = dev->data;
+
+	__ASSERT_NO_MSG(data->vref_mv);
+	return bit * data->vref_mv * config->utcpd.vbvol.vbscale.value / BIT_MASK(10);
```

Comment:
just checking: is this ok on 32 bit for the whole range of expected values or it needs a casting up to 64 to avoid overflowing?

Reply:
cast to `uint64_t` to avoid overflow